test/e2e/framework/framework.go

```diff
@@ -488,6 +497,63 @@ func SendUDP(address, msg string) (string, error) {
 	return string(b[:n]), nil
 }
 
+// SendGameServerTCP sends a message to a gameserver and returns its reply
+// finds the first tcp port from the spec to send the message to,
+// returns error if no Ports were allocated
+func SendGameServerTCP(gs *agonesv1.GameServer, msg string) (string, error) {
+	if len(gs.Status.Ports) == 0 {
+		return "", errors.New("Empty Ports array")
+	}
+
+	// use first tcp port
+	for _, p := range gs.Spec.Ports {
+		if p.Protocol == corev1.ProtocolTCP {
+			return SendGameServerTCPToPort(gs, p.Name, msg)
+		}
+	}
+	return "", errors.New("No UDP ports")
+}
+
+// SendGameServerTCPToPort sends a message to a gameserver at the named port and returns its reply
+// returns error if no Ports were allocated or a port of the specified name doesn't exist
+func SendGameServerTCPToPort(gs *agonesv1.GameServer, portName string, msg string) (string, error) {
+	if len(gs.Status.Ports) == 0 {
+		return "", errors.New("Empty Ports array")
+	}
+	var port agonesv1.GameServerStatusPort
+	for _, p := range gs.Status.Ports {
+		if p.Name == portName {
+			port = p
+		}
+	}
+	address := fmt.Sprintf("%s:%d", gs.Status.Address, port.Port)
+	return SendTCP(address, msg)
+}
+
+// SendTCP sends a message to an address, and returns its reply if
+// it returns one in 30 seconds
+func SendTCP(address, msg string) (string, error) {
+	conn, err := net.Dial("tcp", address)
+	if err != nil {
+		log.Fatalf("Could not Dial: %v", err)
+	}
+
+	err = conn.SetReadDeadline(time.Now().Add(30 * time.Second))
```

Comment:
nit: Should be "cuddled":

```
if err := conn.SetReadDeadline(...); err != nil {
  log.Fatalf(...)
}
```

Reply:
Can you clarify what you mean by "cuddled"?

Reply:
Yes; it's what I put into the comment - you assign the error variable in the if statement to restrict the scope of `err` to just the if block. Then is the convention when the variable isn't needed outside of the if block to keep the variable's scope as small as possible.